server/src/main/scala/org/http4s/server/middleware/GZip.scala

```diff
@@ -63,35 +72,29 @@ object GZip {
     0.toByte)                           // Operating system
   )
 
-  private final case class TrailerGen(crc: CRC32 = new CRC32(), var inputLength: Int = 0)
+  private final class TrailerGen(val crc: CRC32 = new CRC32(), var inputLength: Int = 0)
 
-  private def deflateWithTrailer[F[_]](level: Int, bufferSize: Int): Pipe[F, Byte, Byte] = {
-    val trailer = TrailerGen()
-    pipe.covary[F, Byte, Byte](_.pull(_.await.flatMap(deflateStep(trailer, level, bufferSize)).or(deflateFinish(trailer))))
-  }
+  private def trailer[F[_]](gen: TrailerGen): Pipe[F, Byte, Byte] =
+    pipe.covary[F, Byte, Byte](_.pull(_.await.flatMap(trailerStep(gen))))
 
-  private def deflateStep(trailer: TrailerGen, level: Int, bufferSize: Int): ((Chunk[Byte], Handle[Pure, Byte])) => Pull[Pure, Byte, Handle[Pure, Byte]] = {
+  private def trailerStep(gen: TrailerGen): ((Chunk[Byte], Handle[Pure, Byte])) => Pull[Pure, Byte, Handle[Pure, Byte]] = {
     case (c, h) =>
       val chunkArr = c.toArray
-      trailer.crc.update(chunkArr)
-      trailer.inputLength = trailer.inputLength + chunkArr.length
-      Pull.outputs(deflate(
-        level = level,
-        nowrap = true,
-        bufferSize = bufferSize
-      )(chunk(c))) >> deflateHandle(trailer, level, bufferSize)(h)
+      gen.crc.update(chunkArr)
```

Comment:
I'm a little squeamish here about the side effect happening in the Pure context, but I don't have a broken test to justify my squeamishness.  I'm going to try to break this, but if I can't, I'll approve. :)